### ISSUE TITLE
Random disconnects

### ISSUE THREAD

**Issue by mcorner** [twilio/twilio-video-ios#45]
### Description

Users disconnect from rooms suddenly.  Their bandwidth and connections seem to be ok, and we are having trouble debugging this.

Is there a way to get signaling logs for Room ids?

### Steps to Reproduce

Hard to reproduce, but we have room ids:
RM7b524933415cc8c076a890e6fd83aadf
RMadf65284a9e6124c16d13ac2bf30a7ae

I can probably dig up more.

#### Code

The code is similar to the sample code, I can provide specific parts if needed.

#### Expected Behavior

Not disconnect or receive some callback as to why.

#### Actual Behavior

Disconnects
#### Reproduces How Often

Not often, less than 1% of calls.  Calls are between iOS (and Android) clients and a representative of ours using a browser with the js library.

#### Logs

We see a participant disconnected callback on the web/js side.

### Versions

ios 2.7.0
js 2.0.0-beta4

#### Video iOS SDK

above

#### Xcode

10

#### iOS Version

12.2

#### iOS Device

7 plus

**Comment by mcorner** [comment]
Also: RMb8b5742df58d6d6dedace8510175d8b7

**Comment by ceaglest** [comment]
Hi @mcorner,

I'm spot checked RM7b524933415cc8c076a890e6fd83aadf, and I see that the iOS Participant's signaling connection timed out ~10.5 minutes after the Room was created.

We can look into these Rooms more deeply. However, I should mention right now that it's unlikely that we will fix deep rooted bugs or issues with the current signaling stack in 2.x. We are working on a 3.0.0-beta1 release (scheduled for May, I don't have an exact beta launch date yet), which uses a brand new signaling Client and communicates with globally available signaling Servers. The updated signaling Client is also available today behind an opt-in flag in [twilio-video.js-2.0.0-beta6](https://github.com/twilio/twilio-video.js/blob/master/CHANGELOG.md#200-beta6-march-15-2019) and newer (I noticed you were using beta5).

As for the major version bump to 3.0, it is mostly due to some changes in auth and the fact that the new Servers would not be covered under existing traffic white-lists.

Best,
Chris

**Comment by mcorner** [comment]
Thanks @ceaglest .  When you say "signaling connection timed out", what more can you tell me about this.  The video/audio abruptly terminated.  Would the most likely cause be a network disruption, handoff, or some kind of bug in the signaling stack, a backend problem, or what?  We were seeing very good bandwidth up until that point and it just dropped.  It seems like a somewhat large number to be a network disruption.

We don't believe it is an app crash as we have very few crashes reported in XCode.

Is there additional logging we should be doing to figure this out?

Thanks.

**Comment by ceaglest** [comment]
Hi @mcorner,

> RMadf65284a9e6124c16d13ac2bf30a7ae
> Also, RMb8b5742df58d6d6dedace8510175d8b7

In both cases the JavaScript Participant has timed out. I would suggest trying 2.0.0-beta6+ and enabling the new signaling Client in ConnectOptions. If a JavaScript Participant with the old signaling stack loses its signaling connection it is not possible to reconnect it. This means that any signaling failure using twilio-video.js-2.0.0-beta5 is going to be terminal.

> RM7b524933415cc8c076a890e6fd83aadf

This Room was indeed the iOS Participant timing out.

> When you say "signaling connection timed out", what more can you tell me about this. The video/audio abruptly terminated. Would the most likely cause be a network disruption, handoff, or some kind of bug in the signaling stack, a backend problem, or what?

What I can see from the Room logs so far is that the Participants's signaling session timed out. If the Client detects that the session is expired it will sever the media connection, which I think is why you are seeing media stop unexpectedly (same goes for the Server side). Prior to this point, the iOS app's Room should have entered the reconnecting state.

The Client and Server negotiate a timeout period for the session, and the Client is responsible for sending periodic updates to maintain the session. If these aren't happening it either means:

1. A bug in the Client where it is not sending updates.
2. A bug in the Server where updates don't extend the session.
3. The Client's connection to the Server was severed (reconnecting state entered) long enough for the session to expire.
4. The Server host is gone (exceptional case, and won't be a problem with the new signaling client).

I can look at occurrences of iOS Client timeouts. If possible, do you have access to debug level logs from an iOS Participant that experienced a signaling timeout?

Best,
Chris

**Comment by mcorner** [comment]
Ok, thanks @ceaglest 

We are not properly logging reconnections in our iOS or Android apps, which we will add now.  We might be able to present something in the UI to indicate that it is reconnecting.  But to confirm, both the iOS and Android client libraries should be reconnecting lost signaling connections?

On the js/web side we will try out the new signaling reconnection to see if that helps (or at least log and confirm that this is what is happening to the connections).

We do not have the deeper debug level logs from our iOS/Android clients.  We just have our own logs of callback.  Is it possible to get at deeper logging information without physical access?

**Comment by ceaglest** [comment]
Hello,

> But to confirm, both the iOS and Android client libraries should be reconnecting lost signaling connections?

Correct, that is the intended behavior for mobile SDKs. It's only JavaScript that is not able to reconnect with the old signaling Client.

> We do not have the deeper debug level logs from our iOS/Android clients. We just have our own logs of callback. Is it possible to get at deeper logging information without physical access?

I can get packet captures from the signaling server, but this may not tell the whole story in the way that Çlient side logs would. I can do so for RM7b524933415cc8c076a890e6fd83aadf.

Best,
Chris

**Comment by ceaglest** [comment]
Hi,

After pulling the pcaps for the iOS Participant in RM7b524933415cc8c076a890e6fd83aadf I see that the update messages were sent periodically with the last send ~540 seconds after the session began. After that no more updates are sent and the session ends at the ~628 second mark.

I don't have access to lower layers (TCP, TLS) in the pcap so if its possible to get Client side debug logs I will investigate further.

Best,
Chris

**Comment by mcorner** [comment]
Ok, that lines up:
11:00:14 session starts
11:00:18 recording starts on our end
11:10:45 media goes black in the recording and the and the iOS participant is disconnected from the room (logged on the web/js side).

Since the media was continuing fine, the disconnection was because of the signaling being dead.  It should have restarted. We will be logging this soon, but if we confirm it means we are losing a decent number of calls purely due to signaling problems (even with a good network connection that is supporting media fine).

**Comment by ceaglest** [comment]
Hi @mcorner,

> We will be logging this soon, but if we confirm it means we are losing a decent number of calls purely due to signaling problems (even with a good network connection that is supporting media fine).

I agree. I would highly recommend upgrading on JavaScript when you get the chance. On iOS if you'd be interested in trying 3.0.0-beta1 when its available that would be great. As for iOS 2.x, I will be able to say more once I see Client side logs.

Best,
Chris

**Comment by mcorner** [comment]
How do you enable more logging on the iOS side?

**Comment by ceaglest** [comment]
Debug level should be enough:

```.swift
TwilioVideo.setLogLevel(.debug)
```

**Comment by mcorner** [comment]
Right, but we need to do this remotely since we can't reproduce it.  I don't support there is a way to provide Twilio a custom logger?  Otherwise we are going to have to shim NSLog.

**Comment by mcorner** [comment]
@ceaglest The more I look at it, the harder it looks to capture NSLog and console.log well.  We really would like to have detailed logs from clients since it would help both of us track down problems.  Any chance there is a way to override where the Twilio logs go?  We could then ship them to one of the logging services.

**Comment by ceaglest** [comment]
Hi @mcorner,

Unfortunately we don't have that ability right now. The best you could do with current releases is to redirect the output of NSLog to somewhere else (like a file). It's clear that there are some betterments to make for Twilio's Voice & Video SDKs when it comes to logging support.

Best,
Chris

**Comment by mcorner** [comment]
Would be really helpful to give you the best information.  There are so many things that can't be reproduced locally and consistently.  Someone asked for it in Android as well:
https://github.com/twilio/video-quickstart-android/issues/264

**Comment by ceaglest** [comment]
Hi @mcorner,

I don't have any updates on improved logging yet, but I did want to mention that we have now released [3.0.0-beta1](https://www.twilio.com/docs/video/changelog-twilio-video-ios-version-3x#300-beta1-june-6-2019) which uses a new signaling transport. The goal is to improve reliability, and time to detect connection failures. We've also deployed signaling Servers globally, bringing them closer to your customers.

If you're interested, we have updated our [Swift examples](https://github.com/twilio/video-quickstart-swift/tree/3.0.0-beta) for the 3.0. APIs.

Best,
Chris

**Comment by ceaglest** [comment]
Please let us know if the signaling reliability is improved after upgrading to 3.x SDKs.

**Comment by nickfishman** [comment]
Hey all, sorry to bump an old discussion -- I'm happy to create a new issue for this if you'd prefer.

Are there any new best practices for obtaining log output from the Twilio Video SDK? We're working on an active support case (7123317) related to this library, and we were asked to set the log level as mentioned [above](https://github.com/twilio/twilio-video-ios/issues/45#issuecomment-488814862) and [in the docs](https://www.twilio.com/docs/video/ios#enabling-debug-logging). Unfortunately, in our case the issue only happens to a small subset of our users and we haven't been able to reproduce it ourselves. We do use a third-party library to ship client logs to the cloud, but that only works when you call its loggers directly. It doesn't automatically pull in the output of NSLog or stdout/stderr.

Is there a way to obtain the output of the Twilio Video SDK without overriding NSLog or doing other low-level overrides? This would really help us pin down issues (and hopefully fix the underlying problem for everyone!)

**Comment by nickfishman** [comment]
In case it helps anyone reading this in the future: we chose to follow the approach of intercepting stdout and stderr as outlined in [this post](https://medium.com/@thesaadismail/eavesdropping-on-swifts-print-statements-57f0215efb42) with some tweaks. This can be rather noisy depending on what else is going on in your app. We added a filter for the text `DEBUG:Twilio` in the handler callback, then sent that along to our regular logging library.
